### DISCUSSION TITLE
Dev Server Error: Invalid hook call - when debugging local reactjs component library

### DISCUSSION THREAD

**Discussion by deanmraz** [vercel/next.js#32749] [Help]
https://user-images.githubusercontent.com/2848187/114093064-824fed00-986f-11eb-901f-625e3d17113c.mov

### What version of Next.js are you using?

10.1.3

### What version of Node.js are you using?

12.22.1

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

n/a - justing using `next dev`

### Describe the Bug

Seeing the invalid hook error only on the development server and not the browser

Context 
- Trying to develop component library locally via yarn link
- Configured the webpack aliases for react and react-dom
- Works on browser, does not work on dev server

### Expected Behavior

Local component library hooks should work on the development server.

### To Reproduce

1. Clone git repository https://github.com/deanmraz/nextjs-server-hook-issue.git
2. cd into `nextjs-app` directory and run `yarn install` (use yarn since local link is configured in package.json via yarn api)
3. Run the dev server `yarn dev`
4. Visit http://localhost:3000 in the browser and click `Render Component`. This adds the query param to display the component so its only loaded in the browser and not the development server. 
5. Click on `Change State` to ensure hooks do work and click on `Refresh, ...` to see it still works even on browser refresh.
6. Click on `Server side render, ...` Observe that it does load initially (this is because its still rendering from the browser)
7. Click on `Refresh, ...`
8. Observe the invalid hook error that is coming from the development server.

**Comment by avevlad** [comment]
@deanmraz hello! same problem, maybe you found workaround?

**Comment by deanmraz** [comment]
@avevlad I ended up building a component that wouldn't render the component on the local server side during development. If you build a production version locally or in prod it will render on the server side. Its very hacky but for me until it's fixed. 

```
// DevComponent.tsx 
import React, { useEffect, useState } from 'react';

const Component: React.FC = ({ children }) => {
  const isDevelopment = process.env.NODE_ENV !== 'production';
  const isServer = typeof window === 'undefined';
  return <>{!(isServer && isDevelopment) && children}</>;
};

export default Component;

```

```
// nextjs pages 
import { ReuseableComponent } from '@custom/resuablecomponent`;

<DevComponent>
  <ReuseableComponent />
</DevComponent>

```

**Comment by avevlad** [comment]
@deanmraz thanks!

This worked for me 🎂 (without next-transpile-modules)

```diff

module.exports = {
  webpack: (config, options) => {
+   if (options.isServer) {
+     config.externals = ['react', ...config.externals];
+   }
+
+   config.resolve.alias['react'] = path.resolve(__dirname, '.', 'node_modules', 'react');

    return config
  },
};
```

https://www.npmjs.com/package/next-transpile-modules#i-have-trouble-with-duplicated-dependencies-or-the-invalid-hook-call-error-in-react
